packages/loopring_v3.js/src/types.ts

```diff
@@ -0,0 +1,188 @@
+import BN = require("bn.js");
+import { SparseMerkleTree } from "./SparseMerkleTree";
+
+export enum BlockType {
+    RING_SETTLEMENT = 0,
+    DEPOSIT,
+    ONCHAIN_WITHDRAWAL,
+    OFFCHAIN_WITHDRAWAL,
+    ORDER_CANCELLATION
+  }
+
+  export enum BlockState {
+    NEW = 0,
+    COMMITTED,
+    VERIFIED,
+    FINALIZED,
+  }
+
+  export enum ForgeMode {
+    AUTO_UPGRADABLE = 0,
+    MANUAL_UPGRADABLE,
+    PROXIED,
+    NATIVE,
+  }
+
+  export interface Block {
+    blockIdx: number;
+
+    blockType: BlockType;
+    blockSize: number;
+    blockVersion: number;
+    data: string;
+    offchainData: string;
+
+    operator: string;
+
+    blockState: BlockState;
+
+    merkleRoot: string;
+
+    committedTimestamp: number;
+    verifiedTimestamp?: number;
+    finalizedTimestamp?: number;
+
+    numRequestsProcessed: number;
+
+    totalNumRequestsProcessed: number;
+    totalNumDepositsProccesed: number;
+    totalNumOnchainWithdrawalsProcessed: number;
+    totalNumTradesProccesed: number;
+    totalNumOffchainWithdrawalsProcessed: number;
+    totalNumOrderCancellationsProcessed: number;
+
+    transactionHash: string;
+  }
+
+  export interface Token {
+    tokenID: number;
+    address: string;
+    enabled: boolean;
+  }
+
+  export interface Deposit {
+    blockIdx?: number;
+    requestIdx?: number;
+
+    depositIdx: number;
+    timestamp: number;
+
+    accountID: number;
+    tokenID: number;
+    amount: BN;
+    publicKeyX: string;
+    publicKeyY: string;
+
+    transactionHash: string;
+  }
+
+  export interface OnchainWithdrawal {
+    blockIdx?: number;
+    requestIdx?: number;
+
+    withdrawalIdx: number;
+    timestamp: number;
+
+    accountID: number;
+    tokenID: number;
+    amountRequested: BN;
+    amountWithdrawn?: BN;
+
+    transactionHash: string;
+  }
+
+  export interface Trade {
+    blockIdx: number;
+    requestIdx: number;
+
+    accountIdA: number;
+    orderIdA: number;
+    buyA: boolean;
+    tokenA: number,
+    fillSA: BN;
+    feeA: BN;
+    protocolFeeA: BN;
+    rebateA: BN;
+
+    accountIdB: number;
+    orderIdB: number;
+    buyB: boolean;
+    tokenB: number,
+    fillSB: BN;
+    feeB: BN;
+    protocolFeeB: BN;
+    rebateB: BN;
+  }
+
+  export interface OffchainWithdrawal {
+    blockIdx: number;
+    requestIdx: number;
+
+    accountID: number;
+    tokenID: number;
+    amount: BN;
+    feeTokenID: number;
+    fee: BN;
```

Comment:
Do we need the following fields:
```
  amountRequested: String,
        amountWithdrawn: String,
        transactionHash: String,
```

Reply:
`amountRequested` is the amount that was requested by the user in the on-chain withdrawal request, `amountWithdrawn` is what was actually withdrawn. So I think `amountWithdrawn` is definitely needed, and `amountRequested` I think could be useful to help a user find the withdrawal he requested on-chain (it could be confusing to see a different value than requested if it doesn't really match with `amountWithdrawn` because his balance was lower).

`transactionHash` is the Ethereum transaction hash the request/block was done. So this allows us to link to an Ethereum block explorer to see the actual Ethereum transaction, which I think would be useful.

Reply:
The comment is regarding the OffchainWithdrawal definition, I think it misses the fields you described.

Reply:
For off-chain withdrawals we don't have the `amountRequested` data on-chain because it's not needed (what was requested is not needed to update the Merkle tree, only what was actually withdrawn). The request is also completely off-chain so there is no `transactionHash` which could point to the Ethereum transaction in which the request was made.